invokeai/app/services/shared/sqlite_migrator/sqlite_migrator_common.py

```diff
@@ -0,0 +1,109 @@
+    :meth:`register_post_callback`.
+
+    If a migration has additional dependencies, it is recommended to use functools.partial to provide
+    the dependencies and register the partial as the migration callback.
+    """
+
+    from_version: int = Field(ge=0, strict=True, description="The database version on which this migration may be run")
+    to_version: int = Field(ge=1, strict=True, description="The database version that results from this migration")
+    migrate: MigrateCallback = Field(description="The callback to run to perform the migration")
+    pre_migrate: list[MigrateCallback] = Field(
+        default=[], description="A list of callbacks to run before the migration"
+    )
+    post_migrate: list[MigrateCallback] = Field(
+        default=[], description="A list of callbacks to run after the migration"
+    )
+
+    @model_validator(mode="after")
+    def validate_to_version(self) -> "Migration":
+        if self.to_version <= self.from_version:
+            raise ValueError("to_version must be greater than from_version")
+        return self
+
+    def __hash__(self) -> int:
+        # Callables are not hashable, so we need to implement our own __hash__ function to use this class in a set.
+        return hash((self.from_version, self.to_version))
+
+    def register_pre_callback(self, callback: MigrateCallback) -> None:
+        """Registers a pre-migration callback."""
+        self.pre_migrate.append(callback)
+
+    def register_post_callback(self, callback: MigrateCallback) -> None:
+        """Registers a post-migration callback."""
+        self.post_migrate.append(callback)
+
+
+class MigrationSet:
+    """A set of Migrations. Performs validation during migration registration and provides utility methods."""
+
+    def __init__(self) -> None:
+        self._migrations: set[Migration] = set()
+
+    def register(self, migration: Migration) -> None:
+        """Registers a migration."""
+        if any(m.from_version == migration.from_version for m in self._migrations):
+            raise MigrationVersionError(f"Migration from {migration.from_version} already registered")
+        if any(m.to_version == migration.to_version for m in self._migrations):
+            raise MigrationVersionError(f"Migration to {migration.to_version} already registered")
+        self._migrations.add(migration)
+
+    def get(self, from_version: int) -> Optional[Migration]:
+        """Gets the migration that may be run on the given database version."""
+        # register() ensures that there is only one migration with a given from_version, so this is safe.
+        return next((m for m in self._migrations if m.from_version == from_version), None)
+
+    def validate_migration_chain(self) -> None:
+        """
+        Validates that the migrations form a single chain of migrations from version 0 to the latest version.
+        Raises a MigrationError if there is a problem.
+        """
```

Comment:
Related to my other comment about the motivation behind pre/post migration callbacks: validation doesn't currently account for them - this seems error-prone.

Reply:
A migrate callback does not change the version (or it shouldn't, at least) - the migrator itself does that. This should be safe - am I missing something?

Reply:
Superseded